server/src/main/java/org/opensearch/transport/ClusterConnectionManager.java

```diff
@@ -110,6 +112,22 @@ public void openConnection(DiscoveryNode node, ConnectionProfile connectionProfi
         internalOpenConnection(node, resolvedProfile, listener);
     }
 
+    @Override
+    public Set<DiscoveryNode> getNodesLeftInProgress() {
+        return this.pendingLeft;
+    }
+
+    @Override
+    public void markPendingLefts(List<DiscoveryNode> nodes) {
+        logger.info("marking pending left for nodes: [{}]", nodes);
+        pendingLeft.addAll(nodes);
```

Comment:
this has to probably snapshot the Connection object also from `connectedNodes` that can be later closed.